contracts/early_bird_special/lib.rs

```diff
@@ -0,0 +1,286 @@
+    use button::{
+        ButtonData, ButtonGame, ButtonGameEnvironment, Error, IButtonGame, Result, Score,
+    };
+    use button_token::{BALANCE_OF_SELECTOR, TRANSFER_SELECTOR};
+    use ink_env::Error as InkEnvError;
+    use ink_lang::{
+        codegen::{initialize_contract, EmitEvent},
+        reflect::ContractEventBase,
+    };
+    use ink_prelude::{format, vec::Vec};
+    use ink_storage::traits::SpreadAllocate;
+
+    /// Event type
+    type Event = <EarlyBirdSpecial as ContractEventBase>::Type;
+
+    /// Event emitted when TheButton is created
+    #[ink(event)]
+    #[derive(Debug)]
+    pub struct ButtonCreated {
+        #[ink(topic)]
+        button_token: AccountId,
+        start: BlockNumber,
+        deadline: BlockNumber,
+    }
+
+    /// Event emitted when account is whitelisted to play the game
+    #[ink(event)]
+    #[derive(Debug)]
+    pub struct AccountWhitelisted {
+        #[ink(topic)]
+        player: AccountId,
+    }
+
+    /// Event emitted when account is blacklisted from playing the game
+    #[ink(event)]
+    #[derive(Debug)]
+    pub struct AccountBlacklisted {
+        #[ink(topic)]
+        player: AccountId,
+    }
+
+    /// Event emitted when TheButton is pressed
+    #[ink(event)]
+    #[derive(Debug)]
+    pub struct ButtonPressed {
+        #[ink(topic)]
+        by: AccountId,
+        when: BlockNumber,
+    }
+
+    /// Event emitted when button's death is triggered
+    #[ink(event)]
+    #[derive(Debug)]
+    pub struct ButtonDeath {}
+
+    #[ink(storage)]
+    #[derive(SpreadAllocate)]
+    pub struct EarlyBirdSpecial {
+        data: ButtonData,
+    }
+
+    impl AccessControlled for EarlyBirdSpecial {
+        type ContractError = Error;
+    }
+
+    impl ButtonGame for EarlyBirdSpecial {
+        fn get(&self) -> &ButtonData {
+            &self.data
+        }
+
+        fn get_mut(&mut self) -> &mut ButtonData {
+            &mut self.data
+        }
+
+        fn score(&self, now: BlockNumber) -> Score {
+            let deadline = ButtonGame::deadline(self);
+            deadline - now
+        }
+    }
+
+    // becasue ink! does not allow generics or trait default implementations
+    impl IButtonGame for EarlyBirdSpecial {
+        #[ink(message)]
+        fn is_dead(&self) -> bool {
+            let now = Self::env().block_number();
+            ButtonGame::is_dead(self, now)
+        }
+
+        #[ink(message)]
+        fn press(&mut self) -> Result<()> {
+            let caller = self.env().caller();
+            let now = Self::env().block_number();
+            ButtonGame::press(self, now, caller)?;
+            Self::emit_event(
+                self.env(),
+                Event::ButtonPressed(ButtonPressed {
+                    by: caller,
+                    when: now,
+                }),
+            );
+            Ok(())
+        }
+
+        #[ink(message)]
+        fn death(&self) -> Result<()> {
```

Comment:
I'm not sure when this is supposed to be called and by whom, but my general comment here is that `death` should not trigger the payouts, because:
-- who will call it, who will pay for the gas?
-- given enough players, death will not fit in one block, and we will be screwed.

Whenever a smart contract has a loop over users than it's already a red flag, because of the problems above. So this should be handled differently I believe:
1) death should be only marked with a boolean flag or so,
2) users need to `claim reward` with a transaction, post death.
And while I see that this is not "nice" UX-wise, there is just no alternative that I would see.

Reply:
This is a great comment. Initially the idea was anyone can call it, thus paying the associated costs, but in practice for "fairness" we (aka the foundation) would. However as you rightly point out given a large enough pool of players this tx can never fit into one block. 

Initially I imagined a "batched" death, that always triggers rewards for next `n` accounts - but this does not go around the problem that the entire vector of accounts needs to be read from storage.

In the end the right design is a right design - this is now, per comment, refactored to `claim_reward`, can still be called by anyone (on behalf of the account), such that we can distribute the rewards post-mortem. I guess we should include this change in the wireframes for the UI? (after buttons death, `press` becomes claim_reward or something to that effect)?

Reply:
Awesome, thanks for introducing `claim_reward`! I've looked at the UI and if it was a "one shot" game only then it would make most sense to have `claim` instead of `press` after pressing for the first time, but if we already plan that the game might have rounds, then probably I would place the `claim` button somewhere else. But I'm bad at UI :P